test_rclcpp/test/test_multithreaded.cpp

```diff
@@ -0,0 +1,197 @@
+// Copyright 2015 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0 //
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "gtest/gtest.h"
+
+#include "rclcpp/rclcpp.hpp"
+#include "rclcpp/executors.hpp"
+
+#include "test_rclcpp/msg/u_int32.hpp"
+#include "test_rclcpp/srv/add_two_ints.hpp"
+
+#ifdef RMW_IMPLEMENTATION
+# define CLASSNAME_(NAME, SUFFIX) NAME ## __ ## SUFFIX
+# define CLASSNAME(NAME, SUFFIX) CLASSNAME_(NAME, SUFFIX)
+#else
+# define CLASSNAME(NAME, SUFFIX) NAME
+#endif
+
+static inline void multi_consumer_pub_sub_test(bool intra_process) {
+  rclcpp::init(0, nullptr);
+  std::string node_topic_name = "multi_consumer";
+  if (intra_process) {
+    node_topic_name += "_intra_process";
+  }
+
+  auto node = rclcpp::Node::make_shared(node_topic_name, intra_process);
+  auto pub = node->create_publisher<test_rclcpp::msg::UInt32>(node_topic_name, 10);
+
+  std::vector<rclcpp::Subscription<test_rclcpp::msg::UInt32>::SharedPtr> subscriptions;
+  uint32_t counter = 0;
+
+  auto callback =
+    [&counter, &intra_process](test_rclcpp::msg::UInt32::ConstSharedPtr msg,
+      const rmw_message_info_t & info) -> void
+    {
+      ++counter;
+      printf("callback() %4u with message data %u\n", counter, msg->data);
+      ASSERT_EQ(intra_process, info.from_intra_process);
+    };
+
+  rclcpp::executors::MultiThreadedExecutor executor;
+  // Try to saturate the MultithreadedExecutor's thread pool with subscriptions
+  for (uint32_t i = 0; i < 2*executor.get_number_of_threads(); i++) {
+    auto sub = node->create_subscription<test_rclcpp::msg::UInt32>(node_topic_name, 10, callback);
+    subscriptions.push_back(sub);
+  }
+
+  executor.add_node(node);
+  auto msg = std::make_shared<test_rclcpp::msg::UInt32>();
+
+  // wait a moment for everything to initialize (TODO: fix nondeterministic startup behavior)
+  rclcpp::utilities::sleep_for(1_ms);
+
+  // sanity check that no callbacks have fired
+  executor.spin_once();
```

Comment:
The comment and the expectation here and below both suggest that this is verifying that no callbacks should fire before publishing a message. I can take this check out if you think it is redundant.